pkg/clients/s3/bucketclients/lifecycleConfig.go

```diff
@@ -0,0 +1,135 @@
+}
+
+// Observe checks if the resource exists and if it matches the local configuration
+func (in *LifecycleConfigurationClient) Observe(ctx context.Context, bucket *v1beta1.Bucket) (ResourceStatus, error) {
+	conf, err := in.client.GetBucketLifecycleConfigurationRequest(&awss3.GetBucketLifecycleConfigurationInput{Bucket: aws.String(meta.GetExternalName(bucket))}).Send(ctx)
+	if err != nil {
+		if s3Err, ok := err.(awserr.Error); ok && s3Err.Code() == "NoSuchLifecycleConfiguration" && in.config == nil {
+			return Updated, nil
+		}
+		return NeedsUpdate, errors.Wrap(err, "cannot get bucket lifecycle")
+	}
+
+	if len(conf.Rules) != 0 && in.config == nil {
+		return NeedsDeletion, nil
+	}
+
+	if cmp.Equal(conf.Rules, in.generateConfiguration()) {
+		return Updated, nil
+	}
+	return NeedsUpdate, nil
+}
+
+func (in *LifecycleConfigurationClient) generateConfiguration() []awss3.LifecycleRule {
+	rules := make([]awss3.LifecycleRule, len(in.config.Rules))
+	for i, local := range in.config.Rules {
+		rule := awss3.LifecycleRule{
+			ID:     local.ID,
+			Status: awss3.ExpirationStatus(local.Status),
+		}
+		if local.AbortIncompleteMultipartUpload != nil {
+			rule.AbortIncompleteMultipartUpload = &awss3.AbortIncompleteMultipartUpload{
+				DaysAfterInitiation: local.AbortIncompleteMultipartUpload.DaysAfterInitiation,
+			}
+		}
+		if local.Expiration != nil {
+			rule.Expiration = &awss3.LifecycleExpiration{
+				Date:                      &local.Expiration.Date.Time,
+				Days:                      local.Expiration.Days,
+				ExpiredObjectDeleteMarker: local.Expiration.ExpiredObjectDeleteMarker,
+			}
+		}
+		if local.NoncurrentVersionExpiration != nil {
+			rule.NoncurrentVersionExpiration = &awss3.NoncurrentVersionExpiration{NoncurrentDays: local.NoncurrentVersionExpiration.NoncurrentDays}
+		}
+		if local.NoncurrentVersionTransitions != nil {
+			rule.NoncurrentVersionTransitions = make([]awss3.NoncurrentVersionTransition, len(local.NoncurrentVersionTransitions))
+			for tIndex, transition := range local.NoncurrentVersionTransitions {
+				rule.NoncurrentVersionTransitions[tIndex] = awss3.NoncurrentVersionTransition{
+					NoncurrentDays: transition.NoncurrentDays,
+					StorageClass:   awss3.TransitionStorageClass(transition.StorageClass),
+				}
+			}
+		}
+		if local.Transitions != nil {
+			rule.Transitions = make([]awss3.Transition, len(local.Transitions))
+			for tIndex, transition := range local.Transitions {
+				rule.Transitions[tIndex] = awss3.Transition{
+					Date:         &transition.Date.Time,
+					Days:         transition.Days,
+					StorageClass: awss3.TransitionStorageClass(transition.StorageClass),
+				}
+			}
+		}
+		rules[i] = rule
+	}
+	return rules
+}
+
+// Create sends a request to have resource created on AWS
+func (in *LifecycleConfigurationClient) Create(ctx context.Context, bucket *v1beta1.Bucket) (managed.ExternalUpdate, error) {
+	if in.config == nil {
```

Comment:
We wouldn't need this check if we use `GenerateLifecycleConfiguration(spec *v1beta1.BucketLifecycleConfiguration) *awss3.BucketLifecycleConfiguration` function with the input `bucket` here.